CF: Added alternatives in command line program

AYESDIE opened the pull request:
Issue: #1542 
This is my first Pull Request here.

**Added Alternatives in CLI:**
- **Interpolation**: `AverageInterpolation`,`RegressionInterpolation`,`SimilarityInterpolation`
- **Neighbour Search**: `CosineSearch`,`LMetricSearch(EuclideanSearch)`,`PearsonSearch`

zoq commented:
I just saw your message in #mlpack, we will take a loot at the PR once we have a chance, in some cases this might take more than 24 hours.

AYESDIE commented:
> I just saw your message in #mlpack, we will take a loot at the PR once we have a chance, in some cases this might take more than 24 hours.

Thank you, I will also fix the Styling issues with my code.

zoq commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -142,17 +161,139 @@ void ComputeRecommendations(CFModel* cf,
-    cf->GetRecommendations(numRecs, recommendations, users.row(0).t());
+
+    //  Making All the possible paths
+    if (ns_algo == "Cosine")
```
Please use camel casing for all names; for more information take a look at: https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions.

zoq commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -125,10 +134,20 @@ PARAM_INT_IN("recommendations", "Number of recommendations to generate for each"
                             const size_t numRecs,
                             arma::Mat<size_t>& recommendations)
 {
+  const string ns_algo = CLI::GetParam<string>("neighbour_search");
```
Picky comment, do you think we should use a more descriptive name something like ``neighbourSearchPolicy`` instead of ``ns_algo``. Let me know what you think.

Wenhao-H commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -142,17 +161,139 @@ void ComputeRecommendations(CFModel* cf,
+    //  Making All the possible paths
+    if (ns_algo == "Cosine")
+    {
+        if (iw_algo == "Average")
```
We use two spaces for one tab : https://github.com/mlpack/mlpack/wiki/DesignGuidelines#tabbing.

Wenhao-H commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -142,17 +161,139 @@ void ComputeRecommendations(CFModel* cf,
+        if (iw_algo == "Average")
+        {
+            cf->GetRecommendations<CosineSearch,
+              AverageInterpolation>(numRecs, recommendations,
```
To wrap a line, please use two tabs (4 spaces).
https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping

Wenhao-H commented on src/mlpack/methods/cf/cf_model.hpp:
```diff
@@ -142,15 +146,21 @@ class CFModel
 
   //! Compute recommendations for all users.
+  template<typename NeighbourSearchPolicy,
+          typename InterpolationPolicy>
```
`typename` doesn't align with the one above.

AYESDIE commented:
Thank you for reviewing my code and apologies for not following the Styling Guide.
I updated my syntax as per the feedback, fixed the issues and added information about Weight Interpolation and Neighbor Search in `PROGRAM_INFO`

Wenhao-H commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -71,6 +80,18 @@ PROGRAM_INFO("Collaborative Filtering", "This program performs collaborative "
     " - 'SVDPP' -- SVD++ using a SGD optimizer\n"
+    "\n\n"
+    "The following neighbor search algorithms can be specified via" +
+    " the " + PRINT_PARAM_STRING("neighbor_search") + " parameter:"
```
Picky comment, we could add a new line here for readability.
```suggestion
    " the " + PRINT_PARAM_STRING("neighbor_search") + " parameter:"
    "\n"
```

Wenhao-H commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -71,6 +80,18 @@ PROGRAM_INFO("Collaborative Filtering", "This program performs collaborative "
+    " - 'Pearson'  -- Pearson Search Algorithm\n"
+    "\n\n"
+    "The following weight interpolation algorithms can be specified via" +
+    " the " + PRINT_PARAM_STRING("interpolation") + " parameter:"
```
Same as above.

Wenhao-H commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -166,7 +313,61 @@ void ComputeRMSE(CFModel* cf)
-  cf->Predict(combinations, predictions);
+
+  if (neighborSearchAlgorithm == "Cosine")
+    {
```
Indent errors for the if-else clauses here.

Wenhao-H commented on src/mlpack/methods/cf/cf_model_impl.hpp:
```diff
@@ -34,26 +34,35 @@ void* GetValueVisitor::operator()(CFType<DecompositionPolicy>* c) const
 
-  c->Predict(combinations, predictions);
+  c->template Predict<NeighborSearchPolicy,
+    InterpolationPolicy>(combinations, predictions);
```
Pedantic issue : use 4 spaces to wrap a line.

AYESDIE commented:
Sorry for these styling issues, even after reading the `Styling Guide`, I think it will take time. And also in huge chunks of code, I miss these syntax ssues sometimes. Thanks for reviewing the code :smile:. Once all these issues are over and it is successfully merged (I hope so), I would love to work on other issues in #1542.

AYESDIE commented:
Yes, I would love to have my name on contributors list =]

Wenhao-H commented:
You can add your name by yourself and include that file in this PR.

AYESDIE commented:
> You can add your name by yourself and include that file in this PR.

Added my name in `mlpack/COPYRIGHT.txt`

Wenhao-H commented:
Let's merge this PR in 2-3 days in case anyone might want to have a look at it:)

rcurtin commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -125,10 +148,22 @@ PARAM_INT_IN("recommendations", "Number of recommendations to generate for each"
+    "i", "Average");
+
+PARAM_STRING_IN("neighbor_search", "Algorithm used for neighbor search.",
+    "f", "Euclidean");
```
Why the `f` here?  Also, other mlpack code that lets a user specify a string algorithm usually uses lowercase, like `euclidean`, `cosine`, `pearson`; so I kight suggest that for these options here.

AYESDIE replied:
I was not sure about what to use, so I chose `f` randomly. If there is a specific alphabet associated, I can update it accordingly.

rcurtin replied:
Nothing is arbitrary. :)  This is a one-letter shorthand for the option, so something more like `n` might be more appropriate if it is not already taken.  But I see that both `n` and `N` are taken, so I might suggest `S` for "search".

rcurtin commented on src/mlpack/methods/cf/cf_main.cpp:
```diff
@@ -138,21 +173,135 @@ void ComputeRecommendations(CFModel* cf,
+              << endl;
+
+    //  All possible alternatives for Recommendations
+    if (neighborSearchAlgorithm == "Cosine")
```
There's actually a nicer and more flexible way to do this that will help expand this in the future.  I think that the idea can also be found in `kmeans_main.cpp`.  Basically it looks like this...

```
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  const string& neighborSearchAlgorithm = CLI::GetParam<string>("neighbor_search");

  if (neighborSearxhAlgorithm == "Cosine")
    GetRecommendations<CosineSearch>(c, numRecs, recommendations);
  else if (neighborSearchAlgorithm == "Euclidean")
    GetRecommendations<EuclideanSearch>(c, numRecs, recommendations);
  ...
}

template<typename NeighborSearchType>
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  const string& interpolationAlgorithm = CLI::GetParam<string>("interpolation_algorithm");

  if (interpolationAlgorithm == "Average")
    GetRecommendations<NeighborSearchType, AverageInterpolation>(c, numRecs, recommendations);
  else if ...
}

template<typename NeighborSearchType, typename InterpolationType>
void GetRecommendations(CF* c, const size_t numRecs, arma::Mat<size_t>& recommendations)
{
  c->GetRecommendations<NeighborSearchType, InterpolationType>(numRecs, recommendations);
}
```

You could do the same for `Predict()`.  Sorry for any errors in the idea code---I wrote this on a phone so maybe it is not perfect. :)

AYESDIE replied:
Thanks for taking the time and writing this part, I looked at it and also `kmeans_main.cpp`. I've updated the PR as per it.

AYESDIE commented:
I will take a look at that. In issue #1542, someone else took the task of the `test feature`, so I didn't look at it. But I'll add it myself.

rcurtin commented:
Oh, you're right, the tests were a separate task.  Sorry for my misunderstanding!  I'll post my idea of a test back in #1542 then. :)

@Wenhao-H if you want to merge this now, nothing else from my side---sorry for holding it up a bit. :(

zoq commented:
Don't want to delay the PR as well, but I think it's a good idea to wait for at least some basic tests; not sure if @bikram99 already started with that part or if @AYESDIE is interested to work on them. Let me know what both of you think.

bikram99 commented:
Sorry I was unable to fix the issue after some tries maybe due to lack of understanding .If someone else wants to work on the issue they can
I apologise for the delay 😅

Wenhao-H commented:
Sorry for the late reply. Yeah it would be better if we can have some basic tests before merging it. @AYESDIE Do you want to work on the tests as well?

AYESDIE commented:
Added tests:

- `CFInterpolationAlgorithmBoundTest`
- `CFInterpolationTest`
- `CFNeighborSearchAlgorithmBoundTest`
- `CFNeighborSearchTest`

There was one issue I had here, while doing the `CFInterpolationTest` and `CFNeighborSearchTest`, I was not able to reuse `outputModel` the way it was being shown in several tests above and as shown by @rcurtin as it was throwing SIGABRT error.

```
// The execution of CF algorithm depends on initial random seed.
  mlpack::math::FixedRandomSeed();
  mlpackMain();

  CFModel* outputModel = std::move(CLI::GetParam<CFModel*>("output_model"));
```

 I rather trained the model for each algorithm (which is definately not appropriate).
```
  ResetSettings();

  // Using ~~any~~ algorithm.
  SetInputParam("training", dataset);
  SetInputParam("max_iterations", int(10));
  SetInputParam("query", query);
  SetInputParam("~~any~~", std::string("~~any~~"));
  SetInputParam("recommendations", 5);
```

rcurtin commented:
Looks good to me.  For the memory issue I think that I wrote the original test wrong.  Basically when we call `ResetSettings()`, it deletes any pointers held by CLI (so, `"input_model"` and `"output_model"`) and will delete them multiple times if those two pointers are the same.  The way this is handled in the other CF tests is to avoid calling `ResetSettings()` but instead manually set whether or not inputs were passed.

In any case, here's a test that should work:

```
/**
 * Ensure that using neighbor search algorithm makes a difference.
 */
BOOST_AUTO_TEST_CASE(CFNeighborSearchTest)
{
  mat dataset;
  data::Load("GroupLensSmall.csv", dataset);

  const int querySize = 7;
  Mat<size_t> query = arma::linspace<Mat<size_t>>(0, querySize - 1, querySize);

  // Query with different neighbor search types.
  ResetSettings();

  // Using euclidean neighbor search algorithm.
  SetInputParam("training", dataset);
  SetInputParam("max_iterations", int(10));
  SetInputParam("query", query);
  SetInputParam("neighbor_search", std::string("euclidean"));
  SetInputParam("recommendations", 5);

  mlpackMain();

  const arma::Mat<size_t> output1 = CLI::GetParam<arma::Mat<size_t>>("output");

  BOOST_REQUIRE_EQUAL(output1.n_rows, 5);
  BOOST_REQUIRE_EQUAL(output1.n_cols, 7);

  // Reset passed parameters.
  CLI::GetSingleton().Parameters()["training"].wasPassed = false;
  CLI::GetSingleton().Parameters()["max_iterations"].wasPassed = false;
  CLI::GetSingleton().Parameters()["algorithm"].wasPassed = false;

  // Using cosine neighbor search algorithm.
  SetInputParam("input_model",
      std::move(CLI::GetParam<CFModel*>("output_model")));
  SetInputParam("query", query);
  SetInputParam("neighbor_search", std::string("cosine"));
  SetInputParam("recommendations", 5);

  mlpackMain();

  const arma::Mat<size_t> output2 = CLI::GetParam<arma::Mat<size_t>>("output");

  BOOST_REQUIRE_EQUAL(output2.n_rows, 5);
  BOOST_REQUIRE_EQUAL(output2.n_cols, 7);

  // Using pearson neighbor search algorithm.
  SetInputParam("input_model",
      std::move(CLI::GetParam<CFModel*>("output_model")));
  SetInputParam("query", query);
  SetInputParam("neighbor_search", std::string("pearson"));
  SetInputParam("recommendations", 5);

  mlpackMain();

  const arma::Mat<size_t> output3 = CLI::GetParam<arma::Mat<size_t>>("output");

  BOOST_REQUIRE_EQUAL(output3.n_rows, 5);
  BOOST_REQUIRE_EQUAL(output3.n_cols, 7);

  // The resulting matrices should be different.
  BOOST_REQUIRE(arma::any(arma::vectorise(output1 != output2)));
  BOOST_REQUIRE(arma::any(arma::vectorise(output1 != output3)));
}
```

Let me know if you have any problems with it.  It should be easy to adapt to the interliation test too.

Otherwise everything looks great and once you fix this I think it is entirely ready for merge.  Thanks for the hard work and for implementing the tests. :+1:

AYESDIE commented:
Updated both `CFNeighborSearchTest` and `CFInterpolationTest`. They ran successfully this time.

zoq commented:
Looks like the travis build timed out, let's restart the build.

Wenhao-H commented:
Thanks for the contribution and reviews! I have been busy and couldn't make timely response.
